src/classlibnative/bcltype/number.cpp

```diff
@@ -134,8 +135,191 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
+{
+    // Step 1:
+    // Extract meta data from the input double value.
+    //
+    // Refer to IEEE double precision floating point format.
+    UINT64 realMantissa = 0;
+    int realExponent = 0;
+    UINT32 mantissaHighBitIdx = 0;
+    if (((FPDOUBLE*)&value)->exp > 0)
+    {
+        realMantissa = ((UINT64)(((FPDOUBLE*)&value)->mantHi) << 32) | ((FPDOUBLE*)&value)->mantLo + ((UINT64)1 << 52);
+        realExponent = ((FPDOUBLE*)&value)->exp - 1075;
+        mantissaHighBitIdx = 52;
+    }
+    else
+    {
+        realMantissa = ((UINT64)(((FPDOUBLE*)&value)->mantHi) << 32) | ((FPDOUBLE*)&value)->mantLo;
+        realExponent = -1074;
+        mantissaHighBitIdx = BigNum::LogBase2(realMantissa);
+    }
+
+    // Step 2:
+    // Calculate the first digit exponent. We should estimate the exponent and then verify it later.
+    //
+    // This is an improvement of the estimation in the original paper.
+    // Inspired by http://www.ryanjuckett.com/programming/printing-floating-point-numbers/
+    //
+    // 0.30102999566398119521373889472449 = log10V2
+    // 0.69 = 1 - log10V2 - epsilon (a small number account for drift of floating point multiplication)
+    int firstDigitExponent = (int)(ceil(double((int)mantissaHighBitIdx + realExponent) * 0.30102999566398119521373889472449 - 0.69));
+
+    // Step 3:
+    // Store the input double value in BigNum format.
+    //
+    // To keep the precision, we represent the double value as numertor/denominator.
+    BigNum numerator;
+    BigNum denominator;
+    if (realExponent > 0)
+    {
+        numerator.SetUInt64(realMantissa);
+        BigNum::ShiftLeft(&numerator, realExponent);
+
+        // Explanation:
+        // value = (realMantissa * 2^realExponent) / (1)
+        denominator.SetUInt64(1);
+    }
+    else
+    {
+        // Explanation:
+        // value = (realMantissa * 2^realExponent) / (1)
+        //       = (realMantissa / 2^(-realExponent)
+        numerator.SetUInt64(realMantissa);
+        BigNum::ShiftLeft(1, -realExponent, denominator);
+    }
+
+    if (firstDigitExponent > 0)
+    {
+        BigNum poweredValue;
+        BigNum::Pow10(firstDigitExponent, poweredValue);
+        denominator.Multiply(poweredValue);
+    }
+    else if (firstDigitExponent < 0)
+    {
+        BigNum poweredValue;
+        BigNum::Pow10(-firstDigitExponent, poweredValue);
+        numerator.Multiply(poweredValue);
+    }
+
+    if (BigNum::Compare(numerator, denominator) >= 0)
+    {
+        // The exponent estimation was incorrect.
+        firstDigitExponent += 1;
+    }
+    else
+    {
+        numerator.Multiply(10);
+    }
+
+    *dec = firstDigitExponent - 1;
+
+    BigNum::PrepareHeuristicDivide(&numerator, &denominator);
+
+    // Step 4:
+    // Calculate digits.
+    //
+    // Output digits until reaching the last but one precision or the numerator becomes zero.
+    int digitsNum = 0;
+    int currentDigit = 0;
+    while (true)
```

Comment:
A UINT32 can represent a value of up to 10^9, so you only need `big_int_val % 9` big integer divisions (and therefore iterations), the rest can be regular UINT32 divisions.